### PR TITLE
[BEAM-5690] Fix Zero value issue with GroupByKey/CountByKey in SparkRunner

### PR DESCRIPTION
**Pull request by bmv126**
With CountByKey transform in sparkRunner the expired window is not removed from the timer set.
This causes the empty list to be present as value. 
2019-09-12 21:46:57,384 [Executor task launch worker for task 510] TRACE org.apache.beam.runners.spark.stateful.SparkGroupAlsoByWindowViaWindowSet  - SqlTransform/BeamAggregationRel_45/Group.CombineFieldsByFields/Group.ByFields/GroupByKey: output elements are TimestampedValueInSingleWindow{value=KV{Row:[90001], []}, timestamp=2019-09-12T16:15:59.999Z, window=[2019-09-12T16:15:00.000Z..2019-09-12T16:16:00.000Z), pane=PaneInfo{isFirst=true, isLast=true, timing=ON_TIME, index=0, onTimeIndex=0}}

This causes zero values to be emitted whenever the watermark advances for the key.

Issue can be reproduced using BeamSql (groupByKey) with SparkRunner also.

Fix is provided to check if the timer has crossed the WindowBoundary + lateness, if so the timer is
removed from SparkTimerInternals.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by bmv126** [comment]
R: @jbonofre  R: @kennknowles

**Comment by bmv126** [comment]
@echauchot I have moved the test case to CreateStreamTest.java file.

**Comment by echauchot** [comment]
@bmv126 Sorry I'm very late on that, I'm not very available, I'll take a look on monday or tuesday

**Comment by echauchot** [comment]
@bmv126 currently taking a look at the watermark logic in your PR. Sorry for the delay once again

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java [hunk 1]
```diff
@@ -451,6 +457,47 @@ public void testAdvanceWatermarkEqualToPositiveInfinityThrows() {
...
+            .apply(Count.perKey());
+
+    PAssert.that("Wrong count value ", output)
+        .satisfies(
+            (SerializableFunction<Iterable<KV<Integer, Long>>, Void>)
+                input -> {
+                  for (KV<Integer, Long> element : input) {
+                    if (element.getKey() == 1) {
+                      Long countValue = element.getValue();
+                      assertNotEquals("Count Value is 0 !!!", 0L, countValue.longValue());
```
Comment (echauchot) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java]: As I understood expired timers are not evicted and the fact that they are triggered entails an empty collection as output. But it is not in 100% of cases right, only in some corners cases ? I see no corner case in this test case, there should be 3 value output (one per 3s window, with timestamp 3, 4 and 8). I don't understand how this test ensures that the fix works.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java [hunk 2]
```diff
@@ -451,6 +457,47 @@ public void testAdvanceWatermarkEqualToPositiveInfinityThrows() {
...
+                TimestampedValue.of(KV.of(1, 400L), instant.plus(Duration.standardSeconds(8L))))
+            .advanceNextBatchWatermarkToInfinity();
+
+    PCollection<KV<Integer, Long>> output =
+        p.apply("create kv Source", kvSource)
+            .apply(
+                "window input",
+                Window.<KV<Integer, Long>>into(FixedWindows.of(Duration.standardSeconds(3L)))
+                    .withAllowedLateness(Duration.ZERO))
+            .apply(Count.perKey());
```
Comment (echauchot) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java]: you're lucky :) , Count is implemented in the sdk as Combine.PerKey which is translated in the current spark runner as GBK + Pardo and GBK's translation calls `SparkGroupAlsoByWindowViaWindowSet` which contains your fix. So your fix is called even if you call Count and not GBK. If the runner translated Combine.perKey directly (as the new spark runner) you would not be testing your fix.

Comment (echauchot) [reply runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java]: that is because translation in batch mode and in streaming mode are different: in streaming it uses the sdk impl based on GBK.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java (hunk 1 above)
Comment (echauchot) [reply runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java]: indeed, the test fails if the fix is not present so, fine for me.

**Comment by echauchot** [comment]
@bmv126  fine for me thanks for your work ! Just address the nits and LGTM

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java [hunk 3]
```diff
@@ -338,6 +339,18 @@ public void outputWindowedValue(
...
+                    .filter(
+                        timer ->
+                            timer
+                                .getTimestamp()
+                                .plus(windowingStrategy.getAllowedLateness())
+                                .isBefore(timerInternals.currentInputWatermarkTime()))
+                    .collect(Collectors.toList());
+
+            filteredTimers.forEach(timerInternals::deleteTimer);
+
```
[the remark this answers is not in the document]
Comment (bmv126) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: I have moved the code to a method in LateDataUtils.

**Comment by bmv126** [comment]
@echauchot  I have addressed the review comment, can you have a look.

**Comment by bmv126** [comment]
Run Java PreCommit

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java [hunk 4]
```diff
@@ -338,6 +338,16 @@ public void outputWindowedValue(
...
           if (!outputs.isEmpty() || !stateInternals.getState().isEmpty()) {
+
+            Collection<TimerInternals.TimerData> expiredInternalTimers =
+                LateDataUtils.getExpiredTimers(
+                    timerInternals.getTimers(),
+                    windowingStrategy,
+                    timerInternals.currentInputWatermarkTime());
+
+            // Remove the expired timer from the timerInternals structure
+            expiredInternalTimers.forEach(timerInternals::deleteTimer);
```
Comment (echauchot) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java]: @bmv126 I would prefer that you add a method to LateDataUtils that drops expired timers rather than a method that just lists them.

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Trigerring of the validates runner tests from the PR seems to be broken. I'm running them from my laptop on your PR branch and they pass.
